pkg/roachprod/vm/ibm/provider.go

```diff
@@ -0,0 +1,909 @@
+		}
+
+		// Get the SSH key ID for the region.
+		sshKeyID, err := p.getSshKeyID(l, keyName, region)
+		if err != nil {
+			return nil, errors.Wrapf(err, "unable to get SSH key ID for region %s", region)
+		}
+
+		g.Add(1)
+
+		go func() {
+			defer g.Done()
+
+			vm, err := p.createInstance(l, instanceOptions{
+				vmName: vmName,
+				region: region,
+				zone:   zone,
+				infraOpts: &infraOpts{
+					vpcID:    p.config.regions[region].vpcID,
+					subnetID: p.config.regions[region].zones[zone].subnetID,
+					imageID:  imageID,
+					sshKeyID: sshKeyID,
+				},
+				vmOpts:       opts,
+				providerOpts: providerOpts,
+				tags:         tags,
+			})
+
+			mutex.Lock()
+			defer mutex.Unlock()
+			if err != nil {
+				createErrors = append(
+					createErrors,
+					errors.Wrapf(
+						err,
+						"failed to create instance %s in region %s", vmName, region,
+					),
+				)
+			} else {
+				vms[i] = *vm
+			}
+		}()
+	}
+
+	g.Wait()
+
+	if len(createErrors) > 0 {
+		return nil, errors.Join(createErrors...)
+	}
+
+	return vms, nil
+}
+
+// Delete is part of the vm.Provider interface.
+// It deletes the VMs in the list.
+func (p *Provider) Delete(l *logger.Logger, vms vm.List) error {
+
+	// We don't delete floating IP addresses as they are billed per creation.
+	// We want to keep a pool of unbound floating IP addresses to use for
+	// future instances.
+	// If we want to delete floating IPs, we need to set the flag to true.
+	deleteFloatingIPs := false
+
+	var g errgroup.Group
+	for _, vm := range vms {
+
+		g.Go(func() error {
+
+			zone := vm.Zone
+
+			// We need to query the instance to get the floating IP address
+			// and the attached volumes.
+			instance, err := newInstanceFromCRN(p, vm.ProviderID, deleteFloatingIPs)
+			if err != nil {
+				return errors.Wrapf(err, "failed to get instance from CRN %s", vm.ProviderID)
+			}
+
+			vpcService, err := p.getVpcServiceFromZone(zone)
+			if err != nil {
+				return errors.Wrapf(err, "failed to get VPC service for zone %s", zone)
+			}
+
+			// Start by deleting the floating IP address if any
+			if deleteFloatingIPs {
+				if instance.networkInterface != nil {
+					for _, fip := range instance.networkInterface.NetworkInterface.FloatingIps {
+						_, err = vpcService.DeleteFloatingIP(&vpcv1.DeleteFloatingIPOptions{
+							ID: fip.ID,
+						})
+						if err != nil {
+							return errors.Wrapf(
+								err, "failed to delete floating IP address %s in zone %s",
+								*instance.networkInterface.ID,
+								zone,
+							)
+						}
+					}
+				}
+			}
+
+			// Then finish by deleting the instance
+			_, err = vpcService.DeleteInstance(&vpcv1.DeleteInstanceOptions{
+				ID: instance.instance.ID,
+			})
+			if err != nil {
+				return errors.Wrapf(
+					err, "failed to delete instance %s in zone %s",
+					*instance.instance.ID,
+					zone,
+				)
+			}
+
+			return nil
+		})
+
+	}
+
+	return g.Wait()
+}
+
+func (p *Provider) DeleteCluster(l *logger.Logger, name string) error {
```

Comment:
Do we actually need it? It doesn't really provide a faster path than `Delete`.

Reply:
It is not needed as a fast tracked method to delete, but I added it because it was getting info from the API via a different entrypoint (getting tagged resources rather than listing instances), and it felt like it could maybe one day be useful to cleanup stuff.

I can drop it if it's confusing.